activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -303,6 +303,151 @@ def _select!(*fields) # :nodoc:
       self
     end
 
+    # Add a Common Table Expression (CTE) that you can then reference within another SELECT statement.
+    #
+    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0))
+    #   # WITH posts_with_tags AS (
+    #   #   SELECT * FROM posts WHERE (tags_count > 0)
+    #   # )
+    #   # SELECT * FROM posts
+    #
+    # Once you define Common Table Expression you can use custom `FROM` value or `JOIN` to reference it.
+    #
+    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0).from("posts_with_tags AS posts")
+    #   # WITH posts_with_tags AS (
+    #   #  SELECT * FROM posts WHERE (tags_count > 0)
+    #   # )
+    #   # SELECT * FROM posts_with_tags AS posts
+    #
+    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0).joins("JOIN posts_with_tags ON posts_with_tags.id = posts.id"))
+    #   # WITH posts_with_tags AS (
+    #   #   SELECT * FROM posts WHERE (tags_count > 0)
+    #   # )
+    #   # SELECT * FROM posts JOIN posts_with_tags ON posts_with_tags.id = posts.id
+    #
+    # You can also pass the expression as a `String`.
```

Comment:
If this only accepted Arel, they could pass in an Arel quoted string.

Hoping most people would be able to get this without documentation.

Reply:
> Is this really necessary? If you start dropping to raw sql, is there really lots of value over `find_by_sql`.

Value over `find_by_sql` is that it returns `ActiveRecord::Relation` instead of an Array. My line of thinking was that if you have a complex SQL in a string you should be able to pass it as a CTE expression.

```
COMPLEX_QUERY = <<-SQL
    ... very complex query getting and calculating stuff about posts directly in the database ...
SQL

Post.with(popular_posts: COMPLEX_QUERY).from("popular_posts AS posts")
```

This could lead to security problems so I dropped support for plain strings. It is still possible to pass Arel quoted string though

```
Post.with(popular_posts: Arel.sql(COMPLEX_QUERY)).from("popular_posts AS posts")
```

but didn't document it not to mention Arel API. Not sure how to communicate how we should communicate that properly?

Reply:
I just checked [documentation](https://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-order) for `order` method and it mentions `Arel.sql` so I suppose we can document it here also

<img width="818" alt="Screenshot 2022-06-27 at 10 22 31" src="https://user-images.githubusercontent.com/5148/175894036-1ede1a2b-440a-41f3-914c-bcf65d93e673.png">.